src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,398 @@
+                r = err;
+            }
+        }
+
+        kmod_module_unref(mod);
+    }
+
+    kmod_module_unref_list(modlist);
+    return r;
+}
+
+static int
+kmod_apply_file(struct kmod_ctx *kmod, int fd)
+{
+    int err = 0, ret;
+    struct sol_file_reader *reader;
+    struct sol_str_slice content = SOL_STR_SLICE_EMPTY;
+    struct sol_str_slice *itr;
+    uint16_t idx;
+    struct sol_vector lines;
+
+    reader = sol_file_reader_fd(fd);
+    if (!reader) {
+        SOL_ERR("Could not open config file");
+        return err;
+    }
+
+    content = sol_file_reader_get_all(reader);
+    if (content.len == 0)
+        goto finish;
+
+    lines = sol_util_str_split(content, "\n", 0);
+    SOL_VECTOR_FOREACH_IDX (&lines, itr, idx) {
+        struct sol_str_slice slice;
+
+        slice = sol_str_slice_remove_leading_whitespace(*itr);
+        if (*slice.data == '#' || *slice.data == ';' || *slice.data == '\0')
+            continue;
+
+        slice = sol_str_slice_remove_trailing_whitespace(slice);
+        ret = kmod_apply_value(kmod, slice);
+        if (ret < 0)
+            err = ret;
+    }
+    sol_vector_clear(&lines);
+
+finish:
+    sol_file_reader_close(reader);
+    return err;
+}
+
+static int
+kmod_apply_filename(struct kmod_ctx *kmod, int dfd, const char *file_name)
+{
+    int fd;
+
+    if (*file_name == '/' || dfd < 0)
+        fd = open(file_name, O_RDONLY | O_CLOEXEC);
+    else
+        fd = openat(dfd, file_name, O_RDONLY | O_CLOEXEC);
+
+    if (fd < 0) {
+        SOL_DBG("Could not open: %s", file_name);
+        return 0;
+    }
+
+    return kmod_apply_file(kmod, fd);
+}
+
+static int
+kmod_conf_cmp(const void *data1, const void *data2)
+{
+    return strcmp((const char *)data1, (const char *)data2);
+}
+
+static void
+read_directory_to_vector(DIR *dir, struct sol_ptr_vector *vec)
+{
+    struct dirent ent, *entbuf = NULL;
+    int r;
+
+    r = readdir_r(dir, &ent, &entbuf);
+    while (r == 0 && entbuf) {
+        char *s, *dot = strrchr(ent.d_name, '.');
+
+        if (!dot || !streq(dot, ".conf")) {
+            SOL_DBG("File name does not end in '.conf', ignoring: %s",
+                ent.d_name);
+            goto next;
+        }
+
+        s = strdup(ent.d_name);
+        if (!s) {
+            SOL_WRN("Could not allocate memory to hold dir entry: %s, skipping.",
+                ent.d_name);
+            goto next;
+        }
+
+        sol_ptr_vector_insert_sorted(vec, s, kmod_conf_cmp);
+
+next:
+        r = readdir_r(dir, &ent, &entbuf);
+    }
+}
+
+static void
+kmod_coldplug_find_devices(struct kmod_data *kdata, DIR *dir)
+{
+    struct dirent ent, *entbuf = NULL;
+    int r, dfd;
+
+    dfd = dirfd(dir);
+    if (dfd < 0) {
+        SOL_DBG("Could not obtain directory file descriptor");
+        return;
+    }
+
+    r = readdir_r(dir, &ent, &entbuf);
+    while (r == 0 && entbuf) {
+        struct stat st;
+
+        if (!strcmp(ent.d_name, ".") || !strcmp(ent.d_name, "..")) {
```

Comment:
Sysfs has no hidden files as far as I remember,  but it wouldn't be a problem if so.